exercises/book/A3-traits-generics/mod.md

```diff
@@ -0,0 +1,151 @@
+
+### A3.2.A Defining the type
+Currently, the `LocalStorageVec` `enum` is incomplete. Give it two variants: `Stack` and `Heap`. `Stack` contains two named fields, `buf` and `usize`. `buf` will be the array containing `N` items of type `T`. The `Heap` variant has an unnamed field containing a `Vec<T>`. If you've defined the `LocalStorageVec` variants correctly, running `cargo test` should output something like
+
+```txt
+running 1 test
+test test::it_compiles ... ignored, This test is just to validate the definition of `LocalStorageVec`. If it compiles, all is OK
+
+test result: ok. 0 passed; 0 failed; 1 ignored; 0 measured; 0 filtered out; finished in 0.00s
+```
+
+This test does (and should) not run, but is just there for checking your variant definition.
+
+
+### A4.2.B `impl`-ing `From<Vec<T>`
+
+Uncomment the test `it_from_vecs`, and add an implementation for `From<Vec<T>>` to `LocalStorageVec<T>`. To do so, copy the following code in your `lib.rs` file and replace the `todo!` macro invocation with your code that creates a heap-based `LocalStorageVec` containing the passed `Vec<T>`.
+
+```rust
+impl<T, const N: usize> From<Vec<T>> for LocalStorageVec<T, N> {
+    fn from(v: Vec<T>) -> Self {
+        todo!("Implement me");
+    }
+}
+```
+
+**Question** 
+1. How would you pronounce the first line of the code you just copied in English?*
+
+Run `cargo test` to validate your implementation.
+
+### A3.2.C `impl LocalStorageVec`
+To make the `LocalStorageVec` more useful, we'll add more methods to it. Create an `impl`-block for `LocalStorageVec`. Don't forget to declare and provide the generic paramereters. For now, to make implementations easier, we will add a bound `T`, requiring that it implements `Copy` and `Default`. First off, uncomment the test called `it_constructs`. Make it compile and pass by creating a assosiated function called `new` on `LocalStorageVec` that creates a new, empty `LocalStorageVec` instance without heap allocation.
+
+The next methods we'll implement are `len`, `push`, `pop`, `insert`, `remove` and `clear`:
+- `len` returns the length of the `LocalStorageVec`
+- `push` appends an item to the end of the `LocalStorageVec` and increments its length. Possibly moves the contents to the heap if they no longer fit on the stack.
+- `pop` removes an item from the end of the `LocalStorageVec`, optionally returns it and decrements its length. If the length is 0, `pop` returns `None`
+- `insert` inserts an item at the given index and increments the length of the `LocalStorageVec`
```

Comment:
There are a lot of places where this exercise repeats a certain action (like implementing a trait for `LocalStorageVec`, or adding a method to it); I don't think it is necessary to implement all these methods to achieve the learning objectives; in particular `insert` , `delete` and `clear` can be skipped; after push and pop I think I have understood how adding methods for this enum works.

Reply:
The idea behind this is not only to learn about adding methods and doing generic stuff, but to get some general exercise with Rust syntax and the standard library. I believe ince `push` and `pop` are imlemented, the rest should not take a lot of time